### PR TITLE
Validate custom RPC URL against the claimed chainId before adding it

### PR DESCRIPTION
**Pull request by B0Y3R-AVA**
## Description

**Ticket: [CP-14367](https://ava-labs.atlassian.net/browse/CP-14367)**

Phase 4.2 of the injected-provider modernization plan (§7.4). Adds two layers of validation to `wallet_addEthereumChain` so a dApp can't silently trick the wallet into adopting a malicious RPC endpoint.

**Stacked PR:** base is `boyer/injected-provider-connect-flow` (Phase 1b — #3862). Review/merge #3862 first; GitHub will retarget this onto `main` as the stack lands. This branch was rebased onto the current Phase 1b head and is a single commit.

Before this change, a dApp could call `wallet_addEthereumChain` with any of:
- `http://` (no TLS, sniffable)
- `https://localhost:8545` / `127.0.0.1` (attacker loopback)
- Private-IP RPCs (`10.x`, `192.168.x`, etc.)
- An HTTPS URL that claims to be Ethereum mainnet (chainId 1) but points at attacker-controlled infrastructure

The first three categories now fail synchronously before we even show the approval modal. The last category fails in `approve()` — after the user has tapped Approve, we probe the RPC with `eth_chainId`; if the response doesn't match the declared chainId, we reject the request and do NOT commit the network to Redux.

### Why split sync and async?

The existing handler had a deliberate comment explaining why the async RPC probe was NOT called in `handle()` — dApps like Aave crash / time out waiting for the approval modal if we block on a network call. The split here respects that: synchronous checks (static URL hygiene) run before the modal; the async chainId probe runs after user approval. Best of both worlds — security + UX.

### Summary of changes

- **NEW** `app/services/network/utils/validateCustomRpcUrl.ts` — synchronous validator. Rejects non-HTTPS, localhost/loopback (including IPv4-mapped `[::ffff:7f00:1]`), IPv6 unique-local (`fc00::/7`, `fd00::/8`), IPv6 link-local (`fe80::/10`), and RFC1918 + CGNAT IPv4 ranges.
- **NEW** `app/services/network/utils/validateCustomRpcUrl.test.ts` — 35 tests covering happy path, malformed URLs, non-HTTPS schemes, local hosts, IPv6 private ranges, IPv4 private ranges, and near-miss cases (172.15/172.32/100.63) that should NOT be flagged. (One `it.each` case uses a `javascript:` URL to assert rejection — scoped `eslint-disable no-script-url` documents the intent.)
- **MODIFIED** `app/store/rpc/handlers/chain/wallet_addEthereumChain/wallet_addEthereumChain.ts` — calls `validateCustomRpcUrl` in `handle()`, calls `isValidRPCUrl(chainId, rpcUrl)` in `approve()` before dispatching the network mutation.
- **MODIFIED** `app/store/rpc/handlers/chain/wallet_addEthereumChain/wallet_addEthereumChain.test.ts` — added 3 new tests in `handle()` (http, localhost, private IP rejected) and 1 new test in `approve()` (chainId mismatch rejected, `addCustomNetwork` NOT dispatched). Removed a stale `mockImplementationOnce` that was bleeding across tests.

### Known trust model limitation

The `isValidRPCUrl` probe only catches accidentally-misconfigured or lazily-spoofing RPCs. A motivated attacker can set up a proxy that responds to `eth_chainId` with `0x1` and forwards nothing else. Closing this gap requires an allowlist of known RPC providers cross-checked against chainlist.org — scoped to a follow-up (Phase 4.3).

## Screenshots/Videos

No UI changes. The only user-visible difference: when a dApp sends an invalid RPC URL, the approval either does not open (sync check) or the dApp gets an error after approval (async check, no network written).

## Testing

**Unit tests (re-run after rebase):**
- `yarn test app/services/network/utils/validateCustomRpcUrl.test.ts` — 35/35 ✅
- `yarn test app/store/rpc/handlers/chain/wallet_addEthereumChain/` — 17/17 ✅ (was 13; added 4 new)

**Static (re-run after rebase):**
- `yarn tsc` — ✅ 0 errors
- `yarn lint` (`--max-warnings=0` on touched files) — ✅ 0 errors, 0 warnings

**Manual dev testing steps:**

1. Rebuild, open the in-app browser, navigate to https://chainlist.org.
2. **Happy path:** Find a valid chain you haven't added (Gnosis, Linea). Tap **Add to Wallet** → approve → network added normally.
3. **Sync rejection** (needs a crafted dApp or a modified chainlist fetch that returns an `http://` URL). Expected: no modal appears, dApp receives invalidParams error.
4. **Async rejection** (chainId mismatch): call `wallet_addEthereumChain` with a legitimate HTTPS RPC but lie about the chainId. E.g., pass `chainId: "0x1"` with `rpcUrls: ["https://api.avax.network/ext/bc/C/rpc"]`. Expected: modal appears, user taps Approve, then `approve()` probes the URL, gets 43114 instead of 1, returns invalidParams. Network is NOT added. dApp sees the error.
5. Trigger a Bitrise build and reference it here: _<build link TBD>_
6. Move the ticket into the "Testing" column on Jira.

## Checklist

- [x] I have performed a self-review of my code
- [x] I have verified the code works (unit tests)
- [x] I have added testing steps
- [x] I have added/updated necessary unit tests
- [ ] I have included screenshots / videos of android and ios (no UI change; happy to capture an error-state toast if requested)
- [ ] I have updated the documentation

[CP-14159]: https://ava-labs.atlassian.net/browse/CP-14159?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[CP-14367]: https://ava-labs.atlassian.net/browse/CP-14367?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: packages/core-mobile/app/store/rpc/handlers/chain/wallet_addEthereumChain/wallet_addEthereumChain.ts [hunk 1]
```diff
@@ -59,6 +61,18 @@ class WalletAddEthereumChainHandler
...
+    // IPs). The async chainId probe is deferred until after user approval
+    // because dApps like Aave time out waiting for an approval modal if we
+    // block on a network call here.
+    const urlCheck = validateCustomRpcUrl(rpcUrl)
+    if (!urlCheck.ok) {
+      return {
+        success: false,
+        error: rpcErrors.invalidParams(urlCheck.reason)
+      }
+    }
```
[the remark this answers is not in the document]
Comment (B0Y3R-AVA) [reply packages/core-mobile/app/store/rpc/handlers/chain/wallet_addEthereumChain/wallet_addEthereumChain.ts]: Good call — fixed in fe127b650. Moved the `validateCustomRpcUrl` call below the `chainAlreadyExists` short-circuit, so a request for an already-known chainId returns `{ success: true, value: null }` and keeps its trusted config regardless of the URL the dApp supplies (matching prior behavior). URL validation now only gates chains we're actually about to add. The pre-existing `rpcUrl`-missing check stays above the short-circuit (unchanged). Added a regression test: re-adding a known chain with an `http://` URL still succeeds.

File: packages/core-mobile/app/services/network/utils/validateCustomRpcUrl.ts [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+// IPv6 prefixes (WHATWG-normalized hostnames include brackets) that indicate
+// loopback, unique-local, or link-local addresses. Prefix-matched because
+// the polyfill may produce `[::ffff:7f00:1]`, `[fc00::1]`, etc.
+const LOCAL_IPV6_PREFIXES = [
+  '[::ffff:7f', // IPv4-mapped loopback (127.0.0.0/8 mapped into v6)
+  '[fc', // unique-local fc00::/7
+  '[fd', // unique-local fd00::/8
+  '[fe80:' // link-local fe80::/10
+]
```
[the remark this answers is not in the document]
Comment (B0Y3R-AVA) [reply packages/core-mobile/app/services/network/utils/validateCustomRpcUrl.ts]: Fixed in fe127b650. Replaced the brittle `[::ffff:7f` prefix with numeric handling: `extractMappedIpv4` pulls the embedded IPv4 out of any IPv4-mapped literal — dotted (`::ffff:127.0.0.1`) or WHATWG hex-normalized (`::ffff:7f00:1`) — and runs it through the IPv4 loopback/private checks, so the result no longer depends on the polyfill's serialization. This also closed two real gaps the prefix approach missed: IPv4-mapped **private** ranges (e.g. `::ffff:10.0.0.1` → was allowed) and loopback beyond `.1` (added `127.0.0.0/8`, was only `127.0.0.1` exact). Added tests for `[::ffff:127.0.0.1]`, `[0:0:0:0:0:0:0:1]`, `[::ffff:10.0.0.1]`, `[::ffff:192.168.0.1]`, `[::ffff:a9fe:1]` (rejected) plus `[::ffff:1.1.1.1]` and a public IPv6 (allowed). Note: this polyfill is WHATWG-compliant so it normalizes fully-expanded forms like `[0:0:0:0:0:0:0:1]` → `[::1]` (verified), but the numeric extraction means we no longer rely on that for the mapped-IPv4 case.

File: packages/core-mobile/app/services/network/utils/validateCustomRpcUrl.ts [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+])
+
+// IPv6 prefixes (WHATWG-normalized hostnames include brackets) for unique-local
+// and link-local addresses. IPv4-mapped addresses are handled numerically by
+// extractMappedIpv4 below rather than by prefix, so they can't be bypassed by
+// an alternate serialization (dotted vs hex) of the embedded IPv4.
+const LOCAL_IPV6_PREFIXES = [
+  '[fc', // unique-local fc00::/7
+  '[fd', // unique-local fd00::/8
+  '[fe80:' // link-local fe80::/10
```
Comment (atn4z7) [inline packages/core-mobile/app/services/network/utils/validateCustomRpcUrl.ts]: LOCAL_IPV6_PREFIXES uses '[fe80:', which only matches fe80::/16. The range fe80::/10 covers fe80–febf, so https://[fe9a::1]/rpc, [fea0::1], [feb0::1] etc. pass the check. The doc comment and PR both claim fe80::/10, so the implementation doesn't match the stated intent. Practical risk is low (link-local RPC needs a zone index and isn't really routable), but either the check should cover the full /10 or the comment should be corrected to say /16.

File: packages/core-mobile/app/services/network/utils/validateCustomRpcUrl.ts [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+ * - localhost / loopback (127/8, ::1, IPv4-mapped ::ffff:127.0.0.0/8)
+ * - RFC1918 private IPv4 ranges (10/8, 172.16/12, 192.168/16)
+ * - link-local IPv4 (169.254/16), IPv6 (fe80::/10)
+ * - CGNAT (100.64/10)
+ * - IPv6 unique-local (fc00::/7)
+ *
+ * Requires `react-native-url-polyfill/auto` (loaded in polyfills/index.js).
+ * The stock Hermes `URL` is a regex-based class that does NOT normalize
+ * octal/integer IPv4 encodings (e.g. `0177.0.0.1` → `127.0.0.1`) or IPv6
+ * literals. Without the polyfill this validator becomes silently bypassable.
```
Comment (atn4z7) [inline packages/core-mobile/app/services/network/utils/validateCustomRpcUrl.ts:18]: This comment makes the polyfill normalization the core of the defense ("silently bypassable" without it), but there's no test pinning it — nothing asserts https://2130706433/rpc, https://0x7f000001/rpc, or https://0177.0.0.1/rpc is rejected. Since the whole guarantee rests on the polyfill behaving as assumed at runtime, could we add an "IPv4 encoding bypasses" block next to the IPv6 regression tests (line 118)?

File: packages/core-mobile/app/services/network/utils/validateCustomRpcUrl.ts (hunk 3 above)
Comment (B0Y3R-AVA) [reply packages/core-mobile/app/services/network/utils/validateCustomRpcUrl.ts]: Good catch — fixed in 288f3f466. Replaced the `[fe80:` prefix (which only matched fe80::/16) with `isLocalIpv6()`, covering the full /10 numerically: `/^f[cd]/i` (fc00::/7 unique-local) and `/^fe[89ab][0-9a-f]:/i` (fe80–febf link-local). Added regression tests for `[fe90::1]`, `[fe9a::1]`, `[fea0::1]`, `[feb0::1]`, `[febf::1]` (rejected) plus boundary cases `[fe7f::1]` and `[fec0::1]` (still allowed — confirms we are not over-blocking).

File: packages/core-mobile/app/services/network/utils/validateCustomRpcUrl.ts (hunk 4 above)
Comment (B0Y3R-AVA) [reply packages/core-mobile/app/services/network/utils/validateCustomRpcUrl.ts:18]: Done in 288f3f466. Added an "alternate IPv4 encodings normalize and are rejected" block that pins the polyfill normalization: `2130706433` (decimal), `0x7f000001` (hex) and `0177.0.0.1` (octal) → `127.0.0.1`, plus `0xa000001` → `10.0.0.1` — all rejected. A regression to a non-normalizing URL impl now fails the suite.

File: packages/core-mobile/app/hooks/browser/useEvmInjectedProvider.ts [hunk 5]
```diff
@@ -330,241 +208,160 @@ export function useEvmInjectedProvider(
...
+        appRouter.navigate('/authorizeInjectedDapp')
+
+        // Safety net: if the screen never mounts or calls back within 90s,
+        // reject so the dApp can surface an error instead of hanging.
+        setTimeout(() => {
+          safeReject({
+            code: -32603,
+            message: 'Connect prompt timed out'
+          })
+        }, CONNECT_APPROVAL_TIMEOUT_MS)
```
[the remark this answers is not in the document]
Comment (B0Y3R-AVA) [reply packages/core-mobile/app/hooks/browser/useEvmInjectedProvider.ts:297]: The fallback timer is already cleared on settle — both `safeResolve` and `safeReject` call `clearTimeout(fallbackTimer)`, and `safeReject` is settled-guarded, so an approve/reject clears it immediately rather than leaving it pending for the full 90s. No accumulation across connect prompts.

File: packages/core-mobile/app/new/features/accountSettings/hooks/useConnectedDapps.ts [hunk 6]
```diff
@@ -49,23 +72,28 @@ export const useConnectedDapps = (): {
...
+  const killDapp = async (dapp: Dapp): Promise<void> => {
+    if (dapp.kind === 'wc') {
+      await killSessions([dapp.session])
+      setApprovedWcSessions(current =>
+        current.filter(s => s.topic !== dapp.session.topic)
+      )
+    } else {
+      dispatch(revokePermission({ domain: dapp.domain }))
+    }
   }
```
[the remark this answers is not in the document]
Comment (B0Y3R-AVA) [reply packages/core-mobile/app/new/features/accountSettings/hooks/useConnectedDapps.ts:90]: Fixed in #3862 (5a369d485 — this connect-flow code only shows up in this PR because rpc-validation is stacked on it). `killDapp` now wraps `killSessions` in try/catch and logs, so a teardown failure no longer surfaces as an unhandled rejection at the non-awaiting `onPress` call site.

File: packages/core-mobile/app/new/features/accountSettings/hooks/useConnectedDapps.ts [hunk 7]
```diff
@@ -49,23 +72,28 @@ export const useConnectedDapps = (): {
...
-  const killAllSessions = async (): Promise<void> => {
-    await killSessions(approvedDappsV2)
-    setApprovedDappsV2([])
+  const killAllDapps = async (): Promise<void> => {
+    await killSessions(approvedWcSessions)
+    setApprovedWcSessions([])
+    injectedDomains.forEach(domain => {
+      dispatch(revokePermission({ domain }))
+    })
   }
```
[the remark this answers is not in the document]
Comment (B0Y3R-AVA) [reply packages/core-mobile/app/new/features/accountSettings/hooks/useConnectedDapps.ts:105]: Fixed in #3862 (5a369d485). `killAllDapps` now catches/logs a `killSessions` rejection and still runs the injected-domain `revokePermission` loop afterward, so a WalletConnect teardown failure cannot leave injected dApps connected or the screen stuck.

**Review by atn4z7 — DISMISSED** [review]

**Review by atn4z7 — APPROVED** [review]

**Review by onghwan — APPROVED** [review]
